openwisp_monitoring/monitoring/static/monitoring/js/chart.js

```diff
@@ -16,6 +16,128 @@
+                charts[0].hovertemplate[i] = 'N/A' + '<extra></extra>';
+            }
+            else if (charts[0].y[i] == 0) {
+                convertToAdaptiveBytes(charts, 0, i, 1, 'B');
+            }
+            else if (charts[0].y[i] < 0.001) {
+                convertToAdaptiveBytes(charts, 0, i, 1000000, 'KB');
+            }
+            else if (charts[0].y[i] < 1) {
+                convertToAdaptiveBytes(charts, 0, i, 1000, 'MB');
+            }
+            else if (charts[0].y[i] < 1000) {
+                convertToAdaptiveBytes(charts, 0, i, 1, 'GB');
+            }
+            else {
+                convertToAdaptiveBytes(charts, 0, i, 0.001, 'TB');
+            }
+        }
+
+        for (i=0; i<charts[1].y.length; i++) {
+            if (yRawVal[i] == null) {
+                charts[1].hovertemplate[i] = 'N/A' + '<extra></extra>';
+            }
+            else if (charts[1].y[i] == 0) {
+                convertToAdaptiveBytes(charts, 1, i, 1, 'B');
+            }
+            else if (charts[1].y[i] < 0.001) {
+                convertToAdaptiveBytes(charts, 1, i, 1000000, 'KB');
+            }
+            else if (charts[1].y[i] < 1) {
+                convertToAdaptiveBytes(charts, 1, i, 1000, 'MB');
+            }
+            else if (charts[1].y[i] < 1000) {
+                convertToAdaptiveBytes(charts, 1, i, 1, 'GB');
+            }
+            else {
+                convertToAdaptiveBytes(charts, 1, i, 0.001, 'TB');
+            }
+        }
+
+        for (i=0; i<charts[2].y.length; i++) {
+            if (yRawVal[i] == null) {
+                charts[2].hovertemplate[i] = 'N/A' + '<extra></extra>';
+            }
+            else if (charts[2].y[i] == 0) {
+                convertToAdaptiveBytes(charts, 2, i, 1, 'B');
+            }
+            else if (charts[2].y[i] < 0.001) {
+                convertToAdaptiveBytes(charts, 2, i, 1000000, 'KB');
+            }
+            else if (charts[2].y[i] < 1) {
+                convertToAdaptiveBytes(charts, 2, i, 1000, 'MB');
+            }
+            else if (charts[2].y[i] < 1000) {
+                convertToAdaptiveBytes(charts, 2, i, 1, 'GB');
+            }
+            else {
+                convertToAdaptiveBytes(charts, 2, i, 0.001, 'TB');
+            }
+        }
+    }
+
+    function adaptiveFilterLayout(charts, layout) {
+        var y = charts[0].y, sum = 0, count = 0;
+        for (var i=0; i<y.length; i++) {
+            sum += y[i];
+        }
+        for (i=0; i<y.length; i++) {
+            if (y[i] != 0) {
+                count++;
+            }
+        }
+        var average = sum / count;
+        if (average < 0.001) {
+            layout.yaxis.title = 'KB';
+        }
+        else if (average < 1) {
+            layout.yaxis.title = 'MB';
+        }
+        else if (average < 1000) {
+            layout.yaxis.title = 'GB';
+        }
+        else if (average > 1000) {
+            layout.yaxis.title = 'TB';
+        }
```

Comment:
```suggestion
        else if (average >= 1000) {
            layout.yaxis.title = 'TB';
        }
```

Otherwise, 1TB data would appear in bytes.

Reply:
One observation: isn't this series of if/else which checks the size of the value repeated all over the place?
Why don't we turn that into a function that returns a list with two values: the unit and the scale to use in `convertToAdaptiveBytes`?

That should allow to have code like:

``` 
var scales = getAdaptiveScale(value),
    unit = scales[0],
    scale = scales[1];

layout.yaxis = unit;
```